Floofbot/Modules/Administration.cs

```diff
@@ -506,6 +510,280 @@ public async Task warnlog([Summary("user")] string user)
         }
 
+        [Command("forgive", RunMode = RunMode.Async)]
+        [Summary("Remove a user's warning or user notes")]
+        [RequireContext(ContextType.Guild)]
+        [RequireUserPermission(GuildPermission.BanMembers)]
+        public async Task forgiveUser([Summary("warning/usernote")] string type = "", [Summary("user")] string badUser = "")
+        {
+            IQueryable warnings = null;
+            if (string.IsNullOrEmpty(type) || string.IsNullOrEmpty(badUser) || (!type.ToLower().Equals("warning") && !type.ToLower().Equals("usernote"))) // invalid parameters
+            {
+                Embed embed = CreateDescriptionEmbed($"💾 Usage: `forgive [warning/usernote] [user]`");
+                await SendEmbed(embed);
+                return;
+            }
+            IUser user = resolveUser(badUser);
+
+            ulong uID;
+            if (user != null)
+            {
+                uID = user.Id;
+            }
+            else if (Regex.IsMatch(badUser, @"\d{17,18}")) // not in server but valid user id
+            {
+                uID = Convert.ToUInt64(badUser);
+            }
+            else // user not in server AND not valid user id
+            {
+                await Context.Channel.SendMessageAsync($"⚠️ Could not find user \"{badUser}\"");
+                return;
+            }
+
+            if (type == "warning") // forgive warning
+            {
+                if (!_floofDB.Warnings.AsQueryable().Where(w => w.UserId == uID && w.GuildId == Context.Guild.Id && w.Forgiven == false).Any()) // there are no warnings for this user in this guild
+                {
+                    await Context.Channel.SendMessageAsync("User has no warnings to forgive!");
+                    return;
+                }
+                warnings = _floofDB.Warnings.AsQueryable()
+                    .Where(u => u.UserId == uID && u.GuildId == Context.Guild.Id && u.Forgiven == false)
+                    .OrderByDescending(x => x.DateAdded).Take(10);
+            }
+            else if (type == "usernote") // forgive usernote
+            {
+                if (!_floofDB.UserNotes.AsQueryable().Where(w => w.UserId == uID && w.GuildId == Context.Guild.Id && w.Forgiven == false).Any()) // there are no user notes for this user in this guild
+                {
+                    await Context.Channel.SendMessageAsync("User has no user notes to forgive!");
+                    return;
+                }
+                warnings = _floofDB.UserNotes.AsQueryable()
+                    .Where(u => u.UserId == uID && u.GuildId == Context.Guild.Id && u.Forgiven == false)
+                    .OrderByDescending(x => x.DateAdded).Take(10);
+            }
+
+            EmbedBuilder builder = new EmbedBuilder();
+            builder.Color = ADMIN_COLOR;
+            if (user == null) // no user, just id in database
+                builder.WithTitle($"Warnings for {badUser}");
```

Comment:
The bot will unfortunately now output `Warnings for` even if the moderator wanted to manipulate user notes.

Reply:
This same thing is in lines 574, 707 and 712

Reply:
Added a ternary to fix this in refactoring